fencing/commands.c

```diff
@@ -483,16 +480,26 @@ schedule_stonith_command(async_command_t * cmd, stonith_device_t * device)
         delay_base = delay_max;
     }
     if (delay_max > 0) {
-        cmd->start_delay =
-            ((delay_max != delay_base)?(rand() % (delay_max - delay_base)):0)
-            + delay_base;
-        crm_notice("Delaying %s on %s for %lldms (timeout=%ds, base=%dms, "
-                   "max=%dms)",
-                    cmd->action, device->id, cmd->start_delay, cmd->timeout,
-                    delay_base, delay_max);
-        cmd->delay_id =
-            g_timeout_add(cmd->start_delay, start_delay_helper, cmd);
+        if (device->cib_registered) {
+            cmd->start_delay =
+                ((delay_max != delay_base)?(rand() % (delay_max - delay_base)):0)
+                + delay_base;
+            crm_notice("Delaying %s on %s for %lldms (timeout=%ds, base=%dms, "
+                    "max=%dms)",
+                        cmd->action, device->id, cmd->start_delay, cmd->timeout,
+                        delay_base, delay_max);
+            cmd->delay_id =
+                g_timeout_add(cmd->start_delay, start_delay_helper, cmd);
+        } else {
+            crm_warn("Device '%s' not registered with cib purging "
+                     "delayed operation %s", device->id, cmd->action);
```

Comment:
!cib_registered == not in cib (not necessarily removed from cib)

api_registered == registered via stonith_admin -R (or something else using the C library API)

To complicate things, a device may be registered by both the CIB and the API, and then removed/disabled from just the CIB, in which case it should remain active because the API registration is still valid. Maybe the "if" above needs to be "if cib_registered or api_registered".

Reply:
wouldn't "cib_registered or api_registered" always be true because otherwise there wouldn't be a device?

Reply:
Ah, right ... but in that case, since device exists when we enter this function, we shouldn't purge anything. Maybe purge in stonith_device_remove()? Looking at st_child_done(), is it possible cmd->device_next could point to an already-removed device at that point? Similarly for devices in stonith_fence_get_devices_cb(). Perhaps all pointers into the device table should actually be device IDs, so that we can error out if the device no longer exists.